best-practices/grpc-apis.md

```diff
@@ -0,0 +1,105 @@
+# Evolving gRPC APIs &mdash; Best Practices
+
+This document covers rules and best practices for evolving gRPC APIs to allow
+safe cross-version usage.
+
+## High Level Guidance
+
+The most important takeaway from this document should be that changing protocol
+buffers used across software releases requires careful planning and analysis of
+behavior both in future releases and past releases. Every change to a protobuf
+defintion should include an analysis of how future releases will cope with the
+field if it is not present as well as how past releases will behave given that
+they will be blind to the new field.
+
+## Rules
+
+### R1. Never, _EVER_ Re-use Protcol Buffer Field Numbers
+
+Every protocol buffer field has an identifying number. When
+[encoded](https://developers.google.com/protocol-buffers/docs/encoding) on the
+wire, this number and the field's type are the only information that is actually
+passed. The result is that it is _never_ safe to re-use a field number. Protocol
+buffers provides the `reserved` keyword allowing field numbers to be marked as
+reserved and prevent them from being re-used. Any field number that has been
+used as part of publicly available code (not _just_ releases; this rule applies
+to anything checked into an official repo) should be considered "spent" and
+marked as reserved if the field is later removed.
+
+```proto
+reserved 1, 3;
+
+string name = 2;
+
+int32 size = 4;  // <-- The first available field number is 4.
+```
+
+### R2. Avoid Introducing and Enable a Feature in the same Release
+
+Maintaining downgrade compatibility when launching a new release requires that
+the release not include any features not supported by the previous release. In
+the strictest sense this means that new features cannot be "enabled" until one
+release after they are deemed "working". As a concrete example, when Google
+Compute Engine introduced multi-queue networking we did so one release later
+than when it was "fully supported" in GCE production. This allowed us to safely
+roll-back the release when a bug elsewhere was encountered while not bricking
+VMs that had booted after multi-queue had been enabled.
+
+TODO: Find a concrete example from Kata rather than from an external source.
```

Comment:
In case it helps - I think there is some non-backwards compat change in one of the components between the current [head](https://github.com/kata-containers/runtime/commit/67b5841153424582db620a96eeaf8eb365e1c12e), and the v1.1.0 tags. If I roll the runtime back to 1.1.0 for instance I can not launch docker containers. I've not figured out where the break was though - if anybody knows (and I suspect an agent/proxy relationship maybe), please speak up...